src/vos/ilog.c

```diff
@@ -0,0 +1,1482 @@
+/**
+ * (C) Copyright 2019 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/**
+ * VOS Object/Key incarnation log
+ * vos/ilog.c
+ *
+ * Author: Jeff Olivier <jeffrey.v.olivier@intel.com>
+ */
+#define D_LOGFAC DD_FAC(vos)
+#include <daos/common.h>
+#include <daos_srv/vos.h>
+#include <daos/btree.h>
+#include "vos_internal.h"
+#include "vos_layout.h"
+#include "ilog.h"
+
+#define ILOG_TREE_ORDER 11
+
+enum {
+	ILOG_ITER_NONE,
+	ILOG_ITER_INIT,
+	ILOG_ITER_READY,
+	ILOG_ITER_FINI,
+};
+
+static struct ilog_callbacks ilog_cbs;
+
+/** Register a set of custom callbacks for tx_id */
+void
+ilog_set_callbacks(const struct ilog_callbacks *cbs)
+{
+	ilog_cbs = *cbs;
+}
+
+/** The ilog is split into two parts.   If there is one entry, the ilog
+ *  is embedded into the root df struct.   If not, a b+tree is used.
+ *  The tree is used more like a set where only the key is used.
+ */
```

Comment:
It could very well be but let's not stop progress because something isn't perfect.   The idea was to get the interface update done and then we can refine it.  A btree was the quickest solution.  That said, it is actually pretty good IMO.   Allocations in this implementation are kept to a minimum by using the space already allocated in a tree node.